### PR TITLE
add fori_loop op to all 3 backends

### PR DESCRIPTION
**Pull request by GallagherCommaJack**
`fori_loop` is extremely useful for keeping compile times down in `jax` by defining single layers as loops with a repeated body (instead of unrolling & compiling separately for each block in a sequence).

adding the primitive allows doing this in a low-level way, though eventually something more similar to flax's [`scan`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/_autosummary/flax.linen.scan.html) might be a nicer API for users to consume

### CODE REVIEW

File: keras_core/backend/tensorflow/core.py [hunk 1]
```diff
@@ -163,5 +163,13 @@ def while_loop(
     )
 
 
+def fori_loop(lower, upper, body_fun, init_val):
```
Comment (fchollet) [inline keras_core/backend/tensorflow/core.py:166]: All backend functions should have the same signature across backends.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR. We can add this op. Please add a corresponding Keras op in `ops/core.py` and add unit tests.

File: keras_core/backend/tensorflow/core.py (hunk 1 above)
Comment (GallagherCommaJack) [reply keras_core/backend/tensorflow/core.py:166]: this should be fixed as of most recent commit

File: keras_core/ops/core.py [hunk 2]
```diff
@@ -270,6 +270,30 @@ def stop_gradient(variable):
...
+            self.body_fun,
+            init_val,
+        )
+
+    def compute_output_spec(self, init_val):
+        return KerasTensor(init_val.shape, dtype=init_val.dtype)
+
+
+@keras_core_export("keras_core.ops.fori_loop")
+def fori_loop(lower, upper, body_fun, init_val):
```
Comment (fchollet) [inline keras_core/ops/core.py:293]: This is public-facing, so please add a complete docstring.

File: keras_core/ops/core_test.py [hunk 3]
```diff
@@ -204,6 +204,15 @@ def body(x, y):
         self.assertAllClose(x, np.ones((2, 3)) * 6)
         self.assertAllClose(y, np.ones((3, 2)) * 6)
 
+    def test_fori_loop(self):
```
Comment (fchollet) [inline keras_core/ops/core_test.py:207]: Also test shape inference correctness.

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

**Comment by GallagherCommaJack** [comment]
added shape inference test and a docstring, tests passed

File: keras_core/ops/core.py [hunk 4]
```diff
@@ -293,6 +293,54 @@ def stop_gradient(variable):
...
+    """
+    For loop implementation.
```
Comment (fchollet) [inline keras_core/ops/core.py:318]: Move one-liner to the line above.

File: keras_core/ops/core.py [hunk 5]
```diff
@@ -293,6 +293,54 @@ def stop_gradient(variable):
...
+        upper: The upper bound of the loop variable.
+        body_fun: A callable that represents the loop body. Must take two arguments:
+            the loop variable and the loop state. The loop state should be updated
+            and returned by this function.
+        init_val: The initial value of the loop state.
+
+    Returns:
+        The final state after the loop.
+
+    Examples:
```
Comment (fchollet) [inline keras_core/ops/core.py:331]: Singular (there's only one example)

File: keras_core/ops/core.py [hunk 6]
```diff
@@ -293,6 +293,54 @@ def stop_gradient(variable):
...
+
+    >>> lower = 0
+    >>> upper = 10
+    >>> body_fun = lambda i, s: (i + 1, s + i)
+    >>> init_val = 0
+    >>> keras_core.ops.fori_loop(lower, upper, body_fun, init_val)
+    45
+    """
+
```
Comment (fchollet) [inline keras_core/ops/core.py:340]: Remove blank line

File: keras_core/ops/core_test.py [hunk 7]
```diff
@@ -205,6 +205,23 @@ def body(x, y):
...
+        def body_fun(i, x):
+            return x + i
+
+        initial_value = np.array(0)
+        result = core.fori_loop(0, 10, body_fun, initial_value)
+        self.assertAllClose(result, 45)
+
+    def test_fori_loop_shape_inference(self):
```
Comment (fchollet) [inline keras_core/ops/core_test.py:216]: Move the shape inference to the test class for shape inference (they're all together)

**Review by fchollet — COMMENTED** [review]
Thanks for the updates! Can you also add it to the numpy backend, now that we have one?

File: keras_core/ops/core.py (hunk 4 above)
Comment (GallagherCommaJack) [reply keras_core/ops/core.py:318]: I don't understand...
Do you mean like this?
```
""" For loop implementation.
```

or something else?

File: keras_core/ops/core.py (hunk 5 above)
Comment (GallagherCommaJack) [reply keras_core/ops/core.py:331]: ah good, note that this exact same typo appears on some other ops (e.g. `stop_gradient` directly above)

**Review by fchollet — APPROVED** [review]
LGTM, thanks!

**Comment by GallagherCommaJack** [comment]
the symbolic call for fori_loop could probably use some work, in particular right now it looks like the graph for the loop body is going to get completely ignored? I'm curious what you think the right way to handle it would be.

**Pull request merged**